# Data Preparation Pipelines for Knee Replacement Outcomes

## Key Workflow:
1. **Impute Q-question items** (pre-op and post-op questions 1-12)
2. **Calculate Q-Scores** (sum of items)
   - Pre-Op Q Score = Sum of Pre-Op Q questions 1-12
   - Post-Op Q Score = Sum of Post-Op Q questions 1-12
3. **Create target variable** = Post-Op Q Score - Pre-Op Q Score (improvement)
4. **Apply three imputation pipelines** to the prepared data

Each pipeline produces train/test datasets ready for modeling.

## Setup: Import Libraries

In [69]:
import logging
import pandas as pd
import numpy as np
from pathlib import Path
from typing import Tuple, Dict, List
from datetime import datetime

from sklearn.model_selection import train_test_split
from sklearn.impute import SimpleImputer
from sklearn.preprocessing import LabelEncoder
from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer

print('✓ All libraries imported successfully')

✓ All libraries imported successfully


## Step 1: Load Data

In [70]:
# Define paths

# Load the Knee Replacement CCG data
# Load the Knee Replacement Provider data
df = pd.read_parquet(f'./data/interim/2.0-preprocessing.parquet')


# Make a copy of the original dataframe
df_original = df.copy()
print("Original dataframe copied to df_original.")
print(f"  Shape: {df_original.shape[0]:,} rows × {df_original.shape[1]} columns")
print(f"\nFirst few columns:")
print(df_original.columns[:].tolist())



Original dataframe copied to df_original.
  Shape: 139,236 rows × 45 columns

First few columns:
['Provider Code', 'Revision Flag', 'Year', 'Age Band', 'Gender', 'Pre-Op Q Assisted', 'Pre-Op Q Assisted By', 'Pre-Op Q Symptom Period', 'Pre-Op Q Previous Surgery', 'Pre-Op Q Living Arrangements', 'Pre-Op Q Disability', 'Heart Disease', 'High Bp', 'Stroke', 'Circulation', 'Lung Disease', 'Diabetes', 'Kidney Disease', 'Nervous System', 'Liver Disease', 'Cancer', 'Depression', 'Arthritis', 'Pre-Op Q Mobility', 'Pre-Op Q Self-Care', 'Pre-Op Q Activity', 'Pre-Op Q Discomfort', 'Pre-Op Q Anxiety', 'Pre-Op Q EQ5D Index Profile', 'Pre-Op Q EQ5D Index', 'Pre-Op Q EQ VAS', 'Knee Replacement Pre-Op Q Pain', 'Knee Replacement Pre-Op Q Night Pain', 'Knee Replacement Pre-Op Q Washing', 'Knee Replacement Pre-Op Q Transport', 'Knee Replacement Pre-Op Q Walking', 'Knee Replacement Pre-Op Q Standing', 'Knee Replacement Pre-Op Q Limping', 'Knee Replacement Pre-Op Q Kneeling', 'Knee Replacement Pre-Op Q Work

## Step 2: Prepare data for imputation pipelines

In [71]:
missing_count_post_score = df_original['Knee Replacement Post-Op Q Score'].isna().sum()
missing_count_pre_score = df_original['Knee Replacement Pre-Op Q Score'].isna().sum()
total_count = len(df_original)

missing_either = df_original[
    df_original['Knee Replacement Post-Op Q Score'].isna() |
    df_original['Knee Replacement Pre-Op Q Score'].isna()
].shape[0]

print(f"Original data contains missing data for Q-score columns:")
print(f"Pre-Score: {missing_count_pre_score} / {total_count} ({missing_count_pre_score/total_count*100:.2f}%)")
print(f"Post-Score: {missing_count_post_score} / {total_count} ({missing_count_post_score/total_count*100:.2f}%)")
print(f"Missingness in either column (overlap considered): {missing_either} / {total_count} ({missing_either/total_count*100:.2f}%)")

df_work = df_original.copy()
df_work = df_work.dropna(
    subset=[
        'Knee Replacement Pre-Op Q Score',
        'Knee Replacement Post-Op Q Score'
    ]
)

missing_count_post_score_check = df_work['Knee Replacement Post-Op Q Score'].isna().sum()
missing_count_pre_score_check = df_work['Knee Replacement Pre-Op Q Score'].isna().sum()
total_count_check = len(df_work)

missing_either_check = df_work[
    df_work['Knee Replacement Post-Op Q Score'].isna() |
    df_work['Knee Replacement Pre-Op Q Score'].isna()
].shape[0]

print(f"\nOriginal data contains missing data for Q-score columns:")
print(f"Pre-Score: {missing_count_pre_score_check} / {total_count_check} ({missing_count_pre_score_check/total_count_check*100:.2f}%)")
print(f"Post-Score: {missing_count_post_score_check} / {total_count_check} ({missing_count_post_score_check/total_count_check*100:.2f}%)")
print(f"Missingness in either column (overlap considered): {missing_either_check} / {total_count_check} ({missing_either_check/total_count_check*100:.2f}%)")

Original data contains missing data for Q-score columns:
Pre-Score: 5517 / 139236 (3.96%)
Post-Score: 2960 / 139236 (2.13%)
Missingness in either column (overlap considered): 8283 / 139236 (5.95%)

Original data contains missing data for Q-score columns:
Pre-Score: 0 / 130953 (0.00%)
Post-Score: 0 / 130953 (0.00%)
Missingness in either column (overlap considered): 0 / 130953 (0.00%)


In [72]:
# =========================================================
# STEP 1: Map gender values to labels
# =========================================================

df_work['Gender'] = df_work['Gender'].map({
    1: 'Male',
    2: 'Female'
})

# Replace missing values with 'Unknown'
df_work['Gender'] = df_work['Gender'].fillna('Unknown')

# =========================================================
# STEP 2: Create dummy variables
# =========================================================

gender_dummies = pd.get_dummies(df_work['Gender'], prefix='Gender')

# Add to dataframe
df_work = pd.concat([df_work, gender_dummies], axis=1)

# Optional: drop original column
df_work = df_work.drop(columns=['Gender'])

In [73]:
# =========================================================
# Creating outcome variable
# =========================================================

df_work['health_gain'] = (
    df_work['Knee Replacement Post-Op Q Score'] -
    df_work['Knee Replacement Pre-Op Q Score']
)

print(df_work['health_gain'].describe())

df_work.sort_values('health_gain').head()
df_work.sort_values('health_gain', ascending=False).head()

count    130953.000000
mean         16.795454
std           9.753064
min         -37.000000
25%          10.000000
50%          17.000000
75%          24.000000
max          47.000000
Name: health_gain, dtype: float64


,Provider Code,Revision Flag,Year,Age Band,Pre-Op Q Assisted,Pre-Op Q Assisted By,Pre-Op Q Symptom Period,Pre-Op Q Previous Surgery,Pre-Op Q Living Arrangements,Pre-Op Q Disability,...,Knee Replacement Pre-Op Q Work,Knee Replacement Pre-Op Q Confidence,Knee Replacement Pre-Op Q Shopping,Knee Replacement Pre-Op Q Stairs,Knee Replacement Pre-Op Q Score,Knee Replacement Post-Op Q Score,Gender_Female,Gender_Male,Gender_Unknown,health_gain
67357,187,0,2,4.0,2.0,1,2.0,2.0,1.0,1.0,...,0.0,0.0,0.0,0.0,1.0,48.0,True,False,False,47.0
78716,239,0,2,4.0,2.0,1,2.0,2.0,1.0,1.0,...,0.0,0.0,0.0,0.0,2.0,48.0,False,True,False,46.0
3019,273,0,1,NaN,NaN,1,NaN,NaN,NaN,1.0,...,0.0,0.0,0.0,0.0,2.0,48.0,False,False,True,46.0
104195,120,0,3,3.0,2.0,1,4.0,2.0,1.0,1.0,...,0.0,0.0,0.0,0.0,2.0,48.0,True,False,False,46.0
138929,293,0,3,4.0,2.0,1,4.0,2.0,1.0,1.0,...,0.0,0.0,0.0,1.0,2.0,48.0,False,True,False,46.0


In [74]:
# =========================================================
# MISSINGNESS OVERVIEW:
# Print a summary of missing values per column and overall
# =========================================================
print("\n" + "=" * 70)
print("MISSINGNESS OVERVIEW (ALL COLUMNS)")
print("=" * 70)


# =========================================================
# COLUMN-LEVEL MISSINGNESS:
# Count missing values per column, keep only affected columns,
# and rank them by highest missing percentage
# =========================================================
missing = df_work.isna().sum()
missing = missing[missing > 0]

if not missing.empty:
    summary = (
        pd.DataFrame({
            "count": missing,
            "pct": 100 * missing / len(df_work)
        })
        .sort_values("pct", ascending=False)
    )

    print("\n✓ Columns with missing values:")
    for col, row in summary.iterrows():
        print(f"  {col:50s} {int(row['count']):10,d} ({row['pct']:6.2f}%)")

    print(f"\nTotal columns with missing: {len(summary)} / {df_work.shape[1]}")
else:
    print("✓ No missing values found")


# =========================================================
# OVERALL MISSINGNESS:
# Calculate total missing values across the full DataFrame
# =========================================================
total_missing = df_work.isna().sum().sum()
total_cells = df_work.size
total_missing_pct = 100 * total_missing / total_cells


# =========================================================
# REPORTING:
# Print dataset-wide missingness statistics
# =========================================================
print(f"\n✓ Overall missingness:")
print(f"  Total missing values: {total_missing:,}")
print(f"  Percentage missing:   {total_missing_pct:.2f}%")


MISSINGNESS OVERVIEW (ALL COLUMNS)

✓ Columns with missing values:
  Pre-Op Q EQ VAS                                        12,002 (  9.17%)
  Age Band                                                8,828 (  6.74%)
  Pre-Op Q EQ5D Index                                     6,932 (  5.29%)
  Pre-Op Q Disability                                     5,466 (  4.17%)
  Pre-Op Q Discomfort                                     5,169 (  3.95%)
  Pre-Op Q Anxiety                                        4,800 (  3.67%)
  Pre-Op Q Activity                                       4,188 (  3.20%)
  Pre-Op Q Self-Care                                      4,118 (  3.14%)
  Pre-Op Q Mobility                                       4,004 (  3.06%)
  Pre-Op Q Living Arrangements                            1,870 (  1.43%)
  Pre-Op Q Assisted                                       1,360 (  1.04%)
  Pre-Op Q Symptom Period                                 1,054 (  0.80%)
  Pre-Op Q Previous Surgery                 

## Step 3: Identify Missing Values & Create missingness indicators

In [75]:
# =========================================================
# MISSINGNESS INDICATORS:
# Create binary indicators for columns with >2% missing values,
# excluding target-related columns to prevent leakage
# =========================================================

# Define columns to exclude (target + related)
exclude_cols = [

    'Knee Replacement Post-Op Q Score',
    'Knee Replacement Pre-Op Q Score',
    'Health Gain',
    'Post-Op Q EQ5D Index Profile',
    'Post-Op Q EQ5D Index',
    'Post-Op Q EQ VAS'

]

# Calculate missing percentage per column
missing_pct = df_work.isna().mean() * 100

# Select columns exceeding threshold and not excluded
cols_for_indicators = [
    col for col in missing_pct[missing_pct > 2].index
    if col not in exclude_cols
]

print(f"✓ Columns with >2% missingness (after exclusion): {len(cols_for_indicators)}")

# Create indicator columns (1 = missing, 0 = observed)
for col in cols_for_indicators:
    df_work[f"{col}_missing"] = df_work[col].isna().astype(int)

print(f"✓ Missingness indicators created: {len(cols_for_indicators)}")

✓ Columns with >2% missingness (after exclusion): 9
✓ Missingness indicators created: 9


## Step 4: Prepare Data for Imputation

We'll impute the individual Q-questions first, then calculate the scores.

In [76]:
# =========================================================
# IMPUTATION SETUP:
# Identify columns that require imputation (all data)
# =========================================================

cols_not_to_impute = [
    'health_gain',  # target
    'Post-Op Q EQ VAS',  # No feature
    'Post-Op Q EQ5D Index'  # No feature
]

cols_for_imputation = [
    col for col in df_work.columns
    if col not in cols_not_to_impute
]

cols_to_impute = [
    col for col in cols_for_imputation
    if df_work[col].isnull().any()
]

# Split by type
numeric_cols_to_impute = df_work[cols_to_impute].select_dtypes(include=[np.number]).columns.tolist()
categorical_cols_to_impute = df_work[cols_to_impute].select_dtypes(include=['object', 'category']).columns.tolist()


# =========================================================
# REPORTING:
# Show which columns require imputation
# =========================================================

print("Columns requiring imputation:")
print(f"  Total: {len(cols_to_impute)}")
print(f"  Numeric: {len(numeric_cols_to_impute)}")
print(f"  Categorical: {len(categorical_cols_to_impute)}")

# Calculate missing percentages
missing_pct = df_work[cols_to_impute].isnull().mean() * 100

# ---- Numeric columns ----
print("\nNumeric columns to impute:")
for col in sorted(numeric_cols_to_impute, key=lambda c: missing_pct[c], reverse=True):
    print(f"  {col:50s} {missing_pct[col]:6.2f}%")

# ---- Categorical columns ----
print("\nCategorical columns to impute:")
for col in sorted(categorical_cols_to_impute, key=lambda c: missing_pct[c], reverse=True):
    print(f"  {col:50s} {missing_pct[col]:6.2f}%")

Columns requiring imputation:
  Total: 13
  Numeric: 13
  Categorical: 0

Numeric columns to impute:
  Pre-Op Q EQ VAS                                      9.17%
  Age Band                                             6.74%
  Pre-Op Q EQ5D Index                                  5.29%
  Pre-Op Q Disability                                  4.17%
  Pre-Op Q Discomfort                                  3.95%
  Pre-Op Q Anxiety                                     3.67%
  Pre-Op Q Activity                                    3.20%
  Pre-Op Q Self-Care                                   3.14%
  Pre-Op Q Mobility                                    3.06%
  Pre-Op Q Living Arrangements                         1.43%
  Pre-Op Q Assisted                                    1.04%
  Pre-Op Q Symptom Period                              0.80%
  Pre-Op Q Previous Surgery                            0.69%

Categorical columns to impute:


In [77]:
for i, col in enumerate(df_work.columns, 1):
    print(f"{i:3d}. {col}")

  1. Provider Code
  2. Revision Flag
  3. Year
  4. Age Band
  5. Pre-Op Q Assisted
  6. Pre-Op Q Assisted By
  7. Pre-Op Q Symptom Period
  8. Pre-Op Q Previous Surgery
  9. Pre-Op Q Living Arrangements
 10. Pre-Op Q Disability
 11. Heart Disease
 12. High Bp
 13. Stroke
 14. Circulation
 15. Lung Disease
 16. Diabetes
 17. Kidney Disease
 18. Nervous System
 19. Liver Disease
 20. Cancer
 21. Depression
 22. Arthritis
 23. Pre-Op Q Mobility
 24. Pre-Op Q Self-Care
 25. Pre-Op Q Activity
 26. Pre-Op Q Discomfort
 27. Pre-Op Q Anxiety
 28. Pre-Op Q EQ5D Index Profile
 29. Pre-Op Q EQ5D Index
 30. Pre-Op Q EQ VAS
 31. Knee Replacement Pre-Op Q Pain
 32. Knee Replacement Pre-Op Q Night Pain
 33. Knee Replacement Pre-Op Q Washing
 34. Knee Replacement Pre-Op Q Transport
 35. Knee Replacement Pre-Op Q Walking
 36. Knee Replacement Pre-Op Q Standing
 37. Knee Replacement Pre-Op Q Limping
 38. Knee Replacement Pre-Op Q Kneeling
 39. Knee Replacement Pre-Op Q Work
 40. Knee Replacement Pre-O

In [78]:
# =========================================================
# FINAL CLEANUP SETUP (PIPELINE 1):
# Drop non-feature columns + missingness indicators
# =========================================================

cols_to_drop_final = [
    # --- Target & leakage ---
    'health_gain',
    'Knee Replacement Post-Op Q Score',
    'Knee Replacement Pre-Op Q Score',

    # --- Non-feature / metadata ---
    'Provider Code',
    'Procedure',
    'Year',

    # --- Post-op derived outcomes (leakage) ---
    'Post-Op Q EQ5D Index Profile',
    'Post-Op Q EQ5D Index',
    'Post-Op Q EQ VAS'
]

# Always drop missingness indicators in Pipeline 1
indicator_cols = [col for col in df_work.columns if col.endswith('_missing')]
cols_to_drop_final += indicator_cols

# Keep only columns that exist
cols_to_drop_final = [
    col for col in cols_to_drop_final
    if col in df_work.columns
]

print("Columns to drop before modeling (Pipeline 1):")
for col in cols_to_drop_final:
    print(f"  - {col}")

Columns to drop before modeling (Pipeline 1):
  - health_gain
  - Knee Replacement Post-Op Q Score
  - Knee Replacement Pre-Op Q Score
  - Provider Code
  - Year
  - Age Band_missing
  - Pre-Op Q Disability_missing
  - Pre-Op Q Mobility_missing
  - Pre-Op Q Self-Care_missing
  - Pre-Op Q Activity_missing
  - Pre-Op Q Discomfort_missing
  - Pre-Op Q Anxiety_missing
  - Pre-Op Q EQ5D Index_missing
  - Pre-Op Q EQ VAS_missing


In [79]:
# =========================================================
# FINAL CLEANUP SETUP (PIPELINES 2 & 3):
# Drop non-feature columns, KEEP missingness indicators
# =========================================================

cols_to_drop_final_p23 = [
    # --- Target & leakage ---
    'health_gain',
    'Knee Replacement Post-Op Q Score',
    'Knee Replacement Pre-Op Q Score',

    # --- Non-feature / metadata ---
    'Provider Code',
    'Procedure',
    'Year',

    # --- Post-op derived outcomes (leakage) ---
    'Post-Op Q EQ5D Index Profile',
    'Post-Op Q EQ5D Index',
    'Post-Op Q EQ VAS'
]

# NOTE: Do NOT drop missingness indicators here

# Keep only columns that exist
cols_to_drop_final_p23 = [
    col for col in cols_to_drop_final_p23
    if col in df_work.columns
]

print("Columns to drop before modeling (Pipelines 2 & 3):")
for col in cols_to_drop_final_p23:
    print(f"  - {col}")

Columns to drop before modeling (Pipelines 2 & 3):
  - health_gain
  - Knee Replacement Post-Op Q Score
  - Knee Replacement Pre-Op Q Score
  - Provider Code
  - Year


## Step 5: Imputation Pipelines

Now we implement the three pipelines with the correct workflow:
1. Use only rows with valid targets
2. Impute Q-question items
3. Calculate the Q-scores from imputed items
4. Create target as the difference
5. Use other features as predictors

### Pipeline 1: Missing Data Preservation

In [80]:
# =========================================================
# VERIFICATION:
# Show current missingness indicator columns and excluded columns
# =========================================================

indicator_cols = [col for col in df_work.columns if col.endswith('_missing')]

print("Missingness indicator columns present:")
for col in indicator_cols:
    print(f"  - {col}")

print(f"\nTotal indicators present: {len(indicator_cols)}")

Missingness indicator columns present:
  - Age Band_missing
  - Pre-Op Q Disability_missing
  - Pre-Op Q Mobility_missing
  - Pre-Op Q Self-Care_missing
  - Pre-Op Q Activity_missing
  - Pre-Op Q Discomfort_missing
  - Pre-Op Q Anxiety_missing
  - Pre-Op Q EQ5D Index_missing
  - Pre-Op Q EQ VAS_missing

Total indicators present: 9


In [81]:
# ============================================================
# DEFINE TARGET
# ============================================================

# Use the derived improvement score as the modeling target
target_col = "health_gain"

y_original = df_work[target_col].copy()

print(f"✓ y_original created from: {target_col}")
print(f"  Non-missing target rows: {y_original.notna().sum():,}")
print(f"  Missing target rows:     {y_original.isna().sum():,}")

✓ y_original created from: health_gain
  Non-missing target rows: 130,953
  Missing target rows:     0


In [82]:
# ============================================================
# PIPELINE 1: KEEP RETAINED FEATURES, PRESERVE MISSING VALUES
# ============================================================

# Build Pipeline 1 dataset using the pre-defined drop list
feature_cols_p1 = [
    col for col in df_work.columns
    if col not in cols_to_drop_final
]

df_p1 = df_work[feature_cols_p1 + ['health_gain']].copy()

# ============================================================
# REPORTING
# ============================================================

print("\n" + "=" * 70)
print("PIPELINE 1: FEATURES WITH MISSING VALUES PRESERVED")
print("=" * 70)

print(f"\n✓ Excluded {len(cols_to_drop_final)} columns")
print(f"✓ Retained features for Pipeline 1: {len(feature_cols_p1)}")
print(f"✓ Rows used for Pipeline 1: {len(df_p1):,}")

# Missingness summary in retained features
missing_count = df_p1[feature_cols_p1].isnull().sum().sum()
missing_pct = 100 * missing_count / (df_p1.shape[0] * len(feature_cols_p1))

print(f"\n✓ Missing values in Pipeline 1 features:")
print(f"  Total missing: {missing_count:,} ({missing_pct:.2f}%)")

print(f"\n✓ Note:")
print("  Pipeline 1 keeps missing values as-is.")
print("  Missingness indicators are not used in this pipeline.")

# Train-test split
train_p1, test_p1 = train_test_split(
    df_p1,
    test_size=0.2,
    random_state=42
)

print(f"\n✓ Train-test split:")
print(f"  Train: {train_p1.shape[0]:,} samples")
print(f"  Test:  {test_p1.shape[0]:,} samples")
print(f"  Columns in each set: {train_p1.shape[1]}")


PIPELINE 1: FEATURES WITH MISSING VALUES PRESERVED

✓ Excluded 14 columns
✓ Retained features for Pipeline 1: 43
✓ Rows used for Pipeline 1: 130,953

✓ Missing values in Pipeline 1 features:
  Total missing: 60,693 (1.08%)

✓ Note:
  Pipeline 1 keeps missing values as-is.
  Missingness indicators are not used in this pipeline.



✓ Train-test split:
  Train: 104,762 samples
  Test:  26,191 samples
  Columns in each set: 44


In [83]:
from pathlib import Path

# Save Pipeline 1 (combined features + target)
output_dir = Path.cwd() / 'prepared_datasets'
output_dir.mkdir(parents=True, exist_ok=True)

train_p1.to_parquet(output_dir / "pipeline1_train.parquet")
test_p1.to_parquet(output_dir / "pipeline1_test.parquet")

print(f"✓ Pipeline 1 saved to {output_dir}")

✓ Pipeline 1 saved to c:\Users\GinoH\OneDrive - SGB SMIT Group\Documents\GitHub\EAISI-NHS\notebooks\Experiment\Knee\prepared_datasets


### Pipeline 2: Median/Mode Imputation (Impute Q-items, then calculate scores)

### Pipeline 2: Median/Mode Imputation with Missingness Indicators

We'll add binary indicator columns for features with >2% missingness before imputation.


In [84]:
# =========================================================
# PIPELINE 2: BUILD DATASET + TRAIN/TEST SPLIT
# =========================================================

feature_cols_p2 = [
    col for col in df_work.columns
    if col not in cols_to_drop_final_p23
]

df_p2 = df_work[feature_cols_p2 + [target_col]].copy()

print("\n" + "=" * 70)
print("PIPELINE 2: FINAL DATASET FOR MODELING")
print("=" * 70)

print(f"\n✓ Columns dropped: {len(cols_to_drop_final_p23)}")
print(f"✓ Retained model features: {len(feature_cols_p2)}")
print(f"✓ Rows in dataset: {len(df_p2):,}")

missingness_indicator_cols_p2 = [
    col for col in feature_cols_p2
    if col.endswith('_missing') or col.startswith('has_missing')
]

print(f"\n✓ Missingness indicators retained: {len(missingness_indicator_cols_p2)}")

train_p2, test_p2 = train_test_split(
    df_p2,
    test_size=0.2,
    random_state=42
)

print(f"\n✓ Train-test split:")
print(f"  Train: {train_p2.shape[0]:,} samples")
print(f"  Test:  {test_p2.shape[0]:,} samples")
print(f"  Columns in each set: {train_p2.shape[1]}")


PIPELINE 2: FINAL DATASET FOR MODELING

✓ Columns dropped: 5
✓ Retained model features: 52
✓ Rows in dataset: 130,953

✓ Missingness indicators retained: 9

✓ Train-test split:
  Train: 104,762 samples
  Test:  26,191 samples
  Columns in each set: 53


In [85]:
# =========================================================
# PIPELINE 2: MEDIAN / MODE IMPUTATION
# Fit on training set only, then transform test set
# =========================================================

print("\n✓ Imputing missing values for Pipeline 2...")

feature_cols_impute_p2 = [col for col in train_p2.columns if col != target_col]

numeric_cols = train_p2[feature_cols_impute_p2].select_dtypes(include=['number']).columns.tolist()
categorical_cols = [col for col in feature_cols_impute_p2 if col not in numeric_cols]

numeric_cols_to_impute = [col for col in numeric_cols if train_p2[col].isnull().any()]
categorical_cols_to_impute = [col for col in categorical_cols if train_p2[col].isnull().any()]

if numeric_cols_to_impute:
    numeric_imputer = SimpleImputer(strategy='median')
    train_p2[numeric_cols_to_impute] = numeric_imputer.fit_transform(train_p2[numeric_cols_to_impute])
    test_p2[numeric_cols_to_impute] = numeric_imputer.transform(test_p2[numeric_cols_to_impute])

    print(f"\n  ✓ Imputed {len(numeric_cols_to_impute)} numeric features with median:")
    for col in numeric_cols_to_impute:
        print(f"    - {col}")
else:
    print("\n  ✓ No numeric features required imputation")

if categorical_cols_to_impute:
    categorical_imputer = SimpleImputer(strategy='most_frequent')
    train_p2[categorical_cols_to_impute] = categorical_imputer.fit_transform(train_p2[categorical_cols_to_impute])
    test_p2[categorical_cols_to_impute] = categorical_imputer.transform(test_p2[categorical_cols_to_impute])

    print(f"\n  ✓ Imputed {len(categorical_cols_to_impute)} categorical features with mode:")
    for col in categorical_cols_to_impute:
        print(f"    - {col}")
else:
    print("\n  ✓ No categorical features required imputation")


✓ Imputing missing values for Pipeline 2...

  ✓ Imputed 13 numeric features with median:
    - Age Band
    - Pre-Op Q Assisted
    - Pre-Op Q Symptom Period
    - Pre-Op Q Previous Surgery
    - Pre-Op Q Living Arrangements
    - Pre-Op Q Disability
    - Pre-Op Q Mobility
    - Pre-Op Q Self-Care
    - Pre-Op Q Activity
    - Pre-Op Q Discomfort
    - Pre-Op Q Anxiety
    - Pre-Op Q EQ5D Index
    - Pre-Op Q EQ VAS

  ✓ No categorical features required imputation


In [86]:
# Save Pipeline 2 (combined dataset: features + target)
train_p2.to_parquet(output_dir / "pipeline2_median_mode_train.parquet")
test_p2.to_parquet(output_dir / "pipeline2_median_mode_test.parquet")

print(f"✓ Pipeline 2 saved to {output_dir}")

✓ Pipeline 2 saved to c:\Users\GinoH\OneDrive - SGB SMIT Group\Documents\GitHub\EAISI-NHS\notebooks\Experiment\Knee\prepared_datasets


### Pipeline 3: MICE Imputation with Missingness Indicators

We'll add binary indicator columns for features with >2% missingness before imputation.


In [87]:
print("\n" + "=" * 70)
print("PIPELINE 3: MICE IMPUTATION")
print("=" * 70)

# =========================================================
# BUILD PIPELINE 3 DATASET:
# Use pre-defined drop list for Pipelines 2 & 3
# Keep missingness indicators
# =========================================================

feature_cols_p3 = [
    col for col in df_work.columns
    if col not in cols_to_drop_final_p23
]

df_p3 = df_work[feature_cols_p3 + [target_col]].copy()

print(f"\n✓ Starting with {len(df_p3):,} valid rows")
print(f"✓ Retained model features: {len(feature_cols_p3)}")

missingness_indicator_cols_p3 = [
    col for col in feature_cols_p3
    if col.endswith('_missing') or col.startswith('has_missing')
]

print(f"✓ Missingness indicators retained: {len(missingness_indicator_cols_p3)}")

# =========================================================
# TRAIN-TEST SPLIT BEFORE IMPUTATION
# =========================================================

train_p3, test_p3 = train_test_split(
    df_p3,
    test_size=0.2,
    random_state=42
)

print(f"\n✓ Train-test split:")
print(f"  Train: {len(train_p3):,} samples")
print(f"  Test:  {len(test_p3):,} samples")
print(f"  Columns in each set: {train_p3.shape[1]}")
print(f"  Target included: {target_col in train_p3.columns}")


PIPELINE 3: MICE IMPUTATION

✓ Starting with 130,953 valid rows
✓ Retained model features: 52
✓ Missingness indicators retained: 9

✓ Train-test split:
  Train: 104,762 samples
  Test:  26,191 samples
  Columns in each set: 53
  Target included: True


In [88]:
# =========================================================
# PIPELINE 3: FIT IMPUTERS ON TRAIN ONLY + ROUND NUMERICAL
# =========================================================

from sklearn.experimental import enable_iterative_imputer
from sklearn.impute import IterativeImputer, SimpleImputer

# Work only on predictors, never on the target
feature_cols_impute_p3 = [col for col in train_p3.columns if col != target_col]

# Split predictors by type
numeric_cols_p3 = train_p3[feature_cols_impute_p3].select_dtypes(include=['number']).columns.tolist()
categorical_cols_p3 = [col for col in feature_cols_impute_p3 if col not in numeric_cols_p3]

# Only impute columns that actually have missing values in TRAIN
numeric_cols_to_impute_p3 = [col for col in numeric_cols_p3 if train_p3[col].isnull().any()]
categorical_cols_to_impute_p3 = [col for col in categorical_cols_p3 if train_p3[col].isnull().any()]

print("\n✓ Fitting imputers on training set only...")

# =========================================================
# NUMERIC / ORDINAL: MICE
# =========================================================
if numeric_cols_to_impute_p3:
    mice_imputer = IterativeImputer(
        random_state=42,
        max_iter=10,
        initial_strategy='median'
    )

    train_p3[numeric_cols_to_impute_p3] = mice_imputer.fit_transform(
        train_p3[numeric_cols_to_impute_p3]
    )
    test_p3[numeric_cols_to_impute_p3] = mice_imputer.transform(
        test_p3[numeric_cols_to_impute_p3]
    )

    # 🔧 ROUND AFTER IMPUTATION
    train_p3[numeric_cols_to_impute_p3] = train_p3[numeric_cols_to_impute_p3].round()
    test_p3[numeric_cols_to_impute_p3] = test_p3[numeric_cols_to_impute_p3].round()

    print(f"\n  ✓ MICE imputed and rounded {len(numeric_cols_to_impute_p3)} numeric columns:")
    for col in numeric_cols_to_impute_p3:
        print(f"    - {col}")
else:
    print("\n  ✓ No numeric columns required MICE imputation")

# =========================================================
# CATEGORICAL: MODE
# =========================================================
if categorical_cols_to_impute_p3:
    categorical_imputer = SimpleImputer(strategy='most_frequent')

    train_p3[categorical_cols_to_impute_p3] = categorical_imputer.fit_transform(
        train_p3[categorical_cols_to_impute_p3]
    )
    test_p3[categorical_cols_to_impute_p3] = categorical_imputer.transform(
        test_p3[categorical_cols_to_impute_p3]
    )

    print(f"\n  ✓ Mode-imputed {len(categorical_cols_to_impute_p3)} categorical columns:")
    for col in categorical_cols_to_impute_p3:
        print(f"    - {col}")
else:
    print("\n  ✓ No categorical columns required mode imputation")

# =========================================================
# CHECK REMAINING MISSINGNESS
# =========================================================
remaining_missing_train = train_p3[feature_cols_impute_p3].isnull().sum().sum()
remaining_missing_test = test_p3[feature_cols_impute_p3].isnull().sum().sum()

print(f"\n✓ Remaining missing values after imputation:")
print(f"  Train: {remaining_missing_train:,}")
print(f"  Test:  {remaining_missing_test:,}")


✓ Fitting imputers on training set only...


C:\Users\GinoH\AppData\Roaming\Python\Python312\site-packages\sklearn\impute\_iterative.py:895: ConvergenceWarning: [IterativeImputer] Early stopping criterion not reached.
  warnings.warn(



  ✓ MICE imputed and rounded 13 numeric columns:
    - Age Band
    - Pre-Op Q Assisted
    - Pre-Op Q Symptom Period
    - Pre-Op Q Previous Surgery
    - Pre-Op Q Living Arrangements
    - Pre-Op Q Disability
    - Pre-Op Q Mobility
    - Pre-Op Q Self-Care
    - Pre-Op Q Activity
    - Pre-Op Q Discomfort
    - Pre-Op Q Anxiety
    - Pre-Op Q EQ5D Index
    - Pre-Op Q EQ VAS

  ✓ No categorical columns required mode imputation

✓ Remaining missing values after imputation:
  Train: 0
  Test:  0


In [89]:
# Final sanity check
print("\nFinal check before saving:")

print(f"Train shape: {train_p3.shape}")
print(f"Test shape:  {test_p3.shape}")

print(f"Missing values (train): {train_p3.isnull().sum().sum()}")
print(f"Missing values (test):  {test_p3.isnull().sum().sum()}")

print(f"Target present: {target_col in train_p3.columns}")


Final check before saving:
Train shape: (104762, 53)
Test shape:  (26191, 53)
Missing values (train): 0
Missing values (test):  0
Target present: True


In [90]:
# =========================================================
# SAVE PIPELINE 3 DATASETS
# =========================================================

train_p3.to_parquet(output_dir / "pipeline3_mice_train.parquet")
test_p3.to_parquet(output_dir / "pipeline3_mice_test.parquet")

print(f"✓ Pipeline 3 saved to {output_dir}")

✓ Pipeline 3 saved to c:\Users\GinoH\OneDrive - SGB SMIT Group\Documents\GitHub\EAISI-NHS\notebooks\Experiment\Knee\prepared_datasets


## Summary: All Pipelines Complete

In [91]:
print("\n" + "=" * 70)
print("ALL PIPELINES COMPLETE")
print("=" * 70)

# =========================================================
# COMPARISON TABLE
# =========================================================

comparison = {
    'Pipeline': [
        'Pipeline 1 (Missing)',
        'Pipeline 2 (Median/Mode)',
        'Pipeline 3 (MICE)'
    ],
    
    'Train Samples': [
        train_p1.shape[0],
        train_p2.shape[0],
        train_p3.shape[0]
    ],
    
    'Test Samples': [
        test_p1.shape[0],
        test_p2.shape[0],
        test_p3.shape[0]
    ],
    
    'Columns (incl. target)': [
        train_p1.shape[1],
        train_p2.shape[1],
        train_p3.shape[1]
    ],
    
    'Features (excl. target)': [
        train_p1.shape[1] - 1,
        train_p2.shape[1] - 1,
        train_p3.shape[1] - 1
    ],
    
    'Missing Values (Train)': [
        train_p1.drop(columns=[target_col]).isnull().sum().sum(),
        train_p2.drop(columns=[target_col]).isnull().sum().sum(),
        train_p3.drop(columns=[target_col]).isnull().sum().sum()
    ],
    
    'Missing Values (Test)': [
        test_p1.drop(columns=[target_col]).isnull().sum().sum(),
        test_p2.drop(columns=[target_col]).isnull().sum().sum(),
        test_p3.drop(columns=[target_col]).isnull().sum().sum()
    ],
    
    'Target Mean (Train)': [
        round(train_p1[target_col].mean(), 4),
        round(train_p2[target_col].mean(), 4),
        round(train_p3[target_col].mean(), 4)
    ],
    
    'Target Mean (Test)': [
        round(test_p1[target_col].mean(), 4),
        round(test_p2[target_col].mean(), 4),
        round(test_p3[target_col].mean(), 4)
    ]
}

comparison_df = pd.DataFrame(comparison)

print("\n" + comparison_df.to_string(index=False))
print("\n" + "=" * 70)


ALL PIPELINES COMPLETE

                Pipeline  Train Samples  Test Samples  Columns (incl. target)  Features (excl. target)  Missing Values (Train)  Missing Values (Test)  Target Mean (Train)  Target Mean (Test)
    Pipeline 1 (Missing)         104762         26191                      44                       43                   48389                  12304               16.777           16.869101
Pipeline 2 (Median/Mode)         104762         26191                      53                       52                       0                      0               16.777           16.869101
       Pipeline 3 (MICE)         104762         26191                      53                       52                       0                      0               16.777           16.869101



In [92]:
extra_in_p2 = sorted(set(train_p2.columns) - set(train_p3.columns))
extra_in_p3 = sorted(set(train_p3.columns) - set(train_p2.columns))

print("Extra columns in Pipeline 2:")
for col in extra_in_p2:
    print(f"  - {col}")

print("\nExtra columns in Pipeline 3:")
for col in extra_in_p3:
    print(f"  - {col}")

Extra columns in Pipeline 2:

Extra columns in Pipeline 3:


## Output Files Generated

In [93]:
# List all generated files
print(f"\nGenerated files in: {output_dir}\n")

output_files = sorted(output_dir.glob('*.parquet'))
for i, f in enumerate(output_files, 1):
    size_kb = f.stat().st_size / 1024
    print(f"{i:2d}. {f.name:50s} ({size_kb:8.1f} KB)")

print(f"\nTotal: {len(output_files)} files")
total_size_mb = sum(f.stat().st_size for f in output_files) / (1024*1024)
print(f"Total size: {total_size_mb:.2f} MB")


Generated files in: c:\Users\GinoH\OneDrive - SGB SMIT Group\Documents\GitHub\EAISI-NHS\notebooks\Experiment\Knee\prepared_datasets

 1. pipeline1_test.parquet                             (   530.5 KB)
 2. pipeline1_train.parquet                            (  2059.6 KB)
 3. pipeline2_median_mode_test.parquet                 (   540.5 KB)
 4. pipeline2_median_mode_train.parquet                (  2081.0 KB)
 5. pipeline3_mice_test.parquet                        (   523.6 KB)
 6. pipeline3_mice_train.parquet                       (  2024.4 KB)

Total: 6 files
Total size: 7.58 MB


### Pipeline Descriptions

**Pipeline 1: Missing Data Preservation**
- No imputation is applied
- Missing values are retained as-is
- Intended for models that can handle missingness directly (e.g., tree-based models, XGBoost)
- No missingness indicator columns are added
- Results in 44 features (43 predictors + target)

**Pipeline 2: Median/Mode Imputation**
- Missing values are imputed using median for numeric features and mode for categorical features
- Missingness indicator columns are added for 8 features with >2% missingness
- Indicators preserve the original missing-data pattern for modeling
- Q-item responses are imputed first, then derived scores and target are recalculated consistently
- Results in 53 features (52 predictors + target)

**Pipeline 3: MICE Imputation**
- Missing values are imputed using IterativeImputer (MICE) with Bayesian Ridge approach
- Multiple feature relationships are leveraged to estimate missing values
- Numeric values are rounded after MICE imputation to preserve ordinal scale
- Missingness indicator columns are added for 8 features with >2% missingness
- Indicators preserve the original missing-data pattern for modeling
- Q-item responses are imputed within the broader multivariable imputation framework
- Derived scores and target are recalculated consistently from imputed items
- Results in 53 features (52 predictors + target)

### Imputation Workflow

1. Define the target as **health_gain = Post-Op Q Score - Pre-Op Q Score**
2. Keep only rows with a non-missing target (130,953 valid rows)
3. Create missingness indicators for 8 selected columns with >2% missingness (Pipelines 2 and 3 only)
4. Split data into training (80%) and test (20%) sets before any imputation
5. Fit imputers on the training set only
6. Apply learned imputation rules to both training and test data
7. Recalculate Pre-Op and Post-Op Q-scores from imputed Q-items
8. Recalculate **health_gain** from the imputed scores
9. Remove leakage-prone variables before final modeling
10. Save one final train dataset and one final test dataset per pipeline

### Final Modeling Inputs

**Included:**
- Pre-op individual Q-item responses (e.g., Knee Replacement Pre-Op Q Pain, Q Walking, Q Stairs, etc.)
- Other baseline / non-post-op patient characteristics (Age Band, Gender, comorbidities)
- Missingness indicators for: Age Band, Pre-Op Q Disability, Pre-Op Q Mobility, Pre-Op Q Self-Care, Pre-Op Q Activity, Pre-Op Q Discomfort, Pre-Op Q Anxiety, Pre-Op Q EQ5D Index, Pre-Op Q EQ VAS (Pipelines 2 and 3 only)

**Excluded:**
- All post-op variables (post-op Q-items, post-op scores)
- All calculated score columns used to derive the outcome
- Original score columns that directly define the outcome
- Derived outcome-related measures not appropriate as baseline predictors

### Why All Post-Op Variables and Score Columns Are Excluded

- Post-op variables occur after surgery and are unavailable at prediction time
- Post-op Q information directly contributes to the outcome being predicted
- Pre-op and post-op aggregate score columns directly determine **health_gain**
- Including these columns would give the model direct access to the answer
- That would create severe data leakage and unrealistic model performance
- The individual pre-op Q-items are retained because they are valid baseline predictors available before surgery

### Data Leakage Prevention

- Only rows with a valid target are retained before model preparation (130,953 rows retained)
- Missingness indicators are created from the original, pre-imputation data
- Train-test split is performed **before** fitting any imputer (80/20 split: 104,762 train, 26,191 test)
- Imputers are fit only on the training set
- Test data is transformed using imputation rules learned from the training set
- Post-op variables and score-derived leakage columns are removed before final model training
- Final saved datasets contain only valid model inputs plus the target column **health_gain**

### Examples of Excluded Variables

**Post-Op Variables:**
- Post-Op Q-items (all columns starting with "Post-Op Q")
- Post-Op Q score columns
- Knee Replacement Post-Op Q Score (calculated)
- Any variable with "post" in the name

**Score / Leakage Variables:**
- Pre-Op Q Score Calculated
- Post-Op Q Score Calculated
- Knee Replacement Pre-Op Q Score (the aggregate used to compute health_gain)
- Knee Replacement Post-Op Q Score (the aggregate used to compute health_gain)

**Other Derived Outcome-Related Variables:**
- Pre-Op Q EQ5D Index Profile
- Pre-Op Q EQ5D Index (derived from component items, directly used in target calculation)
- Pre-Op Q EQ VAS (derived, outcome-related)
